### PR TITLE
Define methods using anonymous modules and `method_missing`

### PR DESCRIPTION
**Pull request by kellyredding**
## The Goal

To alter the design to no longer explicitly define raw methods onto objects using NsOptions.

Before root-level namespace methods and root-level proxy methods would be defined directly onto the receiver of NsOptions calls.  In addition, namespaces would define raw methods on themselves as options and sub-namespaces were defined.
## The Solution

Now, root-level namespace and proxy methods are defined on anonymous modules and those anonymous modules are mixed into the receiver.  Namespace instances themselves use `method_missing` magic to expose their options and sub-namespaces.
## The Result

There is no change to the end-user experience.  However, user's will be able to extend NsOptions receivers by overwriting root-level namespaces and proxy methods and calling their `super`.  In addition, users will be able to mix in extensions to these methods.

### CODE REVIEW

**Comment by kellyredding** [comment]
@jcredding - yo!

**Comment by kellyredding** [comment]
@jcredding - btw, the system tests were key here.  Thanks for putting the original work into them - helped me find a few sneaky bugs in the end.

I didn't alter them much except to remove legacy behavior.

**Comment by jcredding** [comment]
@kellyredding - Looked through all the code (besides all the tests), everything looks real good. Nicely done sir.

**Comment by kellyredding** [comment]
@jcredding - thanks!

I just updated the readme and tried to clean it up and make it less verbose (more digestable).  Will you please read over and make sure it still makes sense and conveys enough to be meaningful?  Thanks man.

That README update should be good to cover #32 for now.

File: README.md [hunk 1]
```diff
@@ -108,343 +67,227 @@ module App
...
-User.preferences.home_page = "/home" # you can set options at this level, though I'm not sure why you would
+App.settings.apply({
+  :root => "/path/to/project",
+  :stage => "development"
+  :new_value => 1
+})
+
+App.settings.root    #=> "/path/to/project"
+App.settings.stage   #=> "development"
+App.settings.new_val #=> 1
```
Comment (jcredding) [inline README.md]: @kellyredding - this doesn't match with the `apply` above

File: README.md [hunk 2]
```diff
@@ -108,343 +67,227 @@ module App
...
+settings.opt2 = BetterCustomTypeClass.new(456)
 
-```ruby
-# self referential value
-options(:selfref) do
-  option :something, :default => "123"
-  option :else, :default => Proc.new { self.something }
-end
+# here, type coercion is performed
+# the is the equivalent of doing: `settings.opt2 = MyCustomTypeClass.new(789)`
```
Comment (jcredding) [inline README.md]: @kellyredding - should be 'this is the equivalent...'

File: README.md [hunk 3]
```diff
@@ -108,343 +67,227 @@ module App
...
-Example.stuff.string # => "1", the same as doing String(1)
-Example.stuff.integer = 5.0
-Example.stuff.integer # => 5, this time it's Integer(5.0)
-Example.stuff.float = "5.0"
-Example.stuff.float # => 5.0, same as Float("5.0")
-```
-
-`Symbol`, `Hash` and `Array` work, but ruby doesn't provide built in type casting for these.
+settings do
+  option :opt1, MyCustomTypeClass, :args => lambda{ ['arg 1's value', 'arg 2's value'] }
```
Comment (jcredding) [inline README.md]: @kellyredding - The strings for arg 1 and arg 2 value need to either mix `"` and `'` or escape the inner apostrophes.

**Comment by jcredding** [comment]
@kellyredding - couple minor things on the README. Other then those, looks good and reads well. Doesn't feel like parts are bolted on as they did previously.

**Comment by kellyredding** [comment]
@jcredding - thanks for that proofreading!  merging in all the things now.

**Pull request merged**